tao_compiler/mlir/disc/transforms/shape_utils.cc

```diff
@@ -0,0 +1,1743 @@
+// Copyright 2022 The BladeDISC Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tensorflow/compiler/mlir/disc/transforms/shape_utils.h"
+
+#include <unordered_set>
+#include <utility>
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/lhlo/IR/lhlo_ops.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // TF:llvm-project
+#include "mlir/Dialect/Tensor/IR/Tensor.h"    // TF:llvm-project
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/MLIRContext.h"  // TF:llvm-project
+#include "mlir/IR/OpDefinition.h"
+#include "tensorflow/compiler/mlir/disc/IR/disc_shape_ops.h"
+#include "tensorflow/compiler/mlir/disc/IR/hlo_disc_ops.h"
+
+namespace mlir {
+namespace disc_ral {
+
+// Supports using EquivalenceClasses for Value
+bool operator<(const ValueWrapper& lhs, const ValueWrapper& rhs) {
+  auto lhs_value = lhs.getValue().getAsOpaquePointer();
+  auto rhs_value = rhs.getValue().getAsOpaquePointer();
+  return lhs_value < rhs_value;
+}
+
+// Merge two symbolDim if they are compatible.
+LogicalResult SymbolDim::Merge(SymbolDim* other) {
+  if (!isDynamic() && !other->isDynamic() &&
+      getDimSize() != other->getDimSize())
+    return failure();
+  if (isDynamic()) dimSize_ = other->getDimSize();
+  return success();
+}
+
+Type ShapeAnalysis::getRefinedType(Value value) {
+  SymbolShape* symbolShape = getShape(value);
+
+  // not-shaped type
+  if (!symbolShape) return value.getType();
+
+  auto ty = value.getType().cast<RankedTensorType>();
+  auto dims = symbolShape->getDims();
+  return RankedTensorType::get(dims, ty.getElementType());
+}
+
+LogicalResult ShapeAnalysis::run() {
+  // Make sure only run once.
+  if (initialized) {
+    op_->emitError() << "re-initialized shape analysis";
+    return failure();
+  }
+  initialized = true;
+
+  return buildShapeMap();
+}
+
+LogicalResult ShapeAnalysis::buildShapeMap() {
+  // Do not merge the region loops. We want to compute shape-map, dim-value-map
+  // and decompose one-by-one.
+  for (auto& region : op_->getRegions()) {
+    if (failed(buildRegionShapeMap(&region))) {
+      return failure();
+    }
+  }
+  for (auto& region : op_->getRegions()) {
```

Comment:
An overall question is, what's the essential difference between these 3 phases? Can we walk the graph once and analyze the ops from different perspectives at the same time ( dim_equality, mul_decompose, etc.)

Reply:
To refactor!